pulp_ansible/app/models.py

```diff
@@ -34,30 +40,73 @@ class Meta:
         )
 
 
-class Collection(Content):
-    """
-    A content type representing a Collection.
-    """
+class Collection(Model):
+    """A model representing a Collection."""
 
     TYPE = 'collection'
 
     namespace = models.CharField(max_length=64)
     name = models.CharField(max_length=64)
+
+    deprecated = models.BooleanField(default=False)
+
+    class Meta:
+        unique_together = (
+            'namespace',
+            'name',
+        )
+
+
+class CollectionVersion(Content):
+    """
+    A content type representing a Collection.
+
+    Fields:
+
+        version (models.CharField): Version string.
+        deprecated (models.BooleanField): Indicates that collection
+            version is deprecated (default: False).
+        quality_score: Total version quality score,
+            floating-point number in range [0, 5].
+        metadata: Artifact metadata in JSON format.
+        contents: Contnet items (e.g. roles, modules) provided by
+            collection in JSON format.
+
+    Relations:
+
+        collection: Reference to a collection model.
+    """
+
+    TYPE = 'collection_version'
+
+    collection = models.ForeignKey(
+        Collection,
+        on_delete=models.CASCADE,
+        related_name='versions',
+        null=True
+    )
     version = models.CharField(max_length=128)
 
+    deprecated = models.BooleanField(default=False)
+    quality_score = models.FloatField(null=True)
+    metadata = psql_fields.JSONField(default=dict)
+    contents = psql_fields.JSONField(default=list)
```

Comment:
We support other dbs (e.g. mariadb, mysql, etc) in pulp_ansible. I'm debating about how we can optimally continue to support them but store this data. My first thought is to store this as a json-encoded string in a textfield. Would that work?

Reply:
One option would be to make it more structured altogether and have more fields. Could we get some examples of this data in practice shown?

Reply:
We are planning to use PostgreSQL FTS for search, aren't we?
Even though it can be de-normalized, in Galaxy we perform some filter queries against data stored in JSON fields.

Reply:
So for full text search, we could still make pulp_ansible work for users who choose not to use postgresql. They just won't have full-text search.

It'd be nice to keep support for other databases if possible. If there's no easy way to support other databases, then we can go the postgresql-only route. I just want to check.

Reply:
I'm concerned about extra complexity it may create for models and how to support this in Django.
PostgreSQL FTS works efficient with appropriate (psql-specific) indexes and tsvector fields created in database. 
I wonder if it's even possible to enable \ disable at least on per-model level in django database specific models.

Reply:
I'd like to at least get a picture of what that complexity would look like before we decide. I'm not ruling out dropping other dbs but I'd like to have all the info before making a decision.

Reply:
The full text search feature is going to raise an exception when the 'q' object is used. So running on MariaDB with that is still feasible with that feature missing. With this change however the whole thing will be postgresql. That is what we talked about at the call so this isn't surprising to me. Today I was thinking about its impact though, and this makes a lot of sense for storing the data formats Galaxy already has, postgreSQL only could make running this with the 10+ other plugins that run on mariaDB could be a problem. It's not a deal-breaker, but I want us to at least consider options.

Can we see more examples of data that goes in there?

Reply:
@bmbouter I think the main question is how to split and maintain driver-specific models \ fields within the same django application?